File name '[name].ts' differs from '[name].d.ts' only in casing when using --build

**TypeScript Version:**  3.1.0-dev.201xxxxx

**Search Terms:**
build
forceConsistentCasingInFileNames
3.0

**Code**

Example project: https://github.com/nicojs/typescript-3-build-demo

After `npm install`, use `npm run build` (or `npx tsc --build src/server`).

**Expected behavior:**

Compile without error.

**Actual behavior:**

Results in error: 

> src/server/main.ts:2:26 - error TS1149: File name 'C:/z/github/nicojs/typescript-3-build-demo/src/shared/Customer.ts' differs from already included file name 'C:/z/github/nicojs/typescript-3-build-demo/dist/shared/Customer.d.ts' only in casing.


**Related Issues:** None that I could find that address this issue.

--- a/src/compiler/program.ts
+++ b/src/compiler/program.ts
@@ -2008,9 +2008,17 @@ namespace ts {
             if (filesByName.has(path)) {
                 const file = filesByName.get(path);
                 // try to check if we've already seen this file but with a different casing in path
-                // NOTE: this only makes sense for case-insensitive file systems
-                if (file && options.forceConsistentCasingInFileNames && getNormalizedAbsolutePath(file.fileName, currentDirectory) !== getNormalizedAbsolutePath(fileName, currentDirectory)) {
-                    reportFileNamesDifferOnlyInCasingError(fileName, file.fileName, refFile, refPos, refEnd);
+                // NOTE: this only makes sense for case-insensitive file systems, and only on files which are not redirected
+                if (file && options.forceConsistentCasingInFileNames) {
+                    let inputName = fileName;
+                    const checkedName = file.fileName;
+                    const isRedirect = toPath(checkedName) !== toPath(inputName);
+                    if (isRedirect) {
+                        inputName = getProjectReferenceRedirect(fileName) || fileName;
+                    }
+                    if (getNormalizedAbsolutePath(checkedName, currentDirectory) !== getNormalizedAbsolutePath(inputName, currentDirectory)) {
+                        reportFileNamesDifferOnlyInCasingError(inputName, checkedName, refFile, refPos, refEnd);
+                    }
                 }
 
                 // If the file was previously found via a node_modules search, but is now being processed as a root file,

--- a/tests/projects/sample1/logic/index.ts
+++ b/tests/projects/sample1/logic/index.ts
@@ -2,3 +2,5 @@ import * as c from '../core/index';
 export function getSecondsInDay() {
     return c.multiply(10, 15);
 }
+import * as mod from '../core/anotherModule';
+export const m = mod;

--- a/tests/projects/sample1/logic/tsconfig.json
+++ b/tests/projects/sample1/logic/tsconfig.json
@@ -1,7 +1,8 @@
 {
     "compilerOptions": {
         "composite": true,
-        "declaration": true
+        "declaration": true,
+        "forceConsistentCasingInFileNames": true
     },
     "references": [
         { "path": "../core" }

--- a/tests/projects/sample1/tests/index.ts
+++ b/tests/projects/sample1/tests/index.ts
@@ -3,3 +3,6 @@ import * as logic from '../logic/index';
 
 c.leftPad("", 10);
 logic.getSecondsInDay();
+
+import * as mod from '../core/anotherModule';
+export const m = mod;

--- a/tests/projects/sample1/tests/tsconfig.json
+++ b/tests/projects/sample1/tests/tsconfig.json
@@ -3,5 +3,10 @@
         { "path": "../core" },
         { "path": "../logic" }
     ],
-    "files": ["index.ts"]
+    "files": ["index.ts"],
+    "compilerOptions": {
+        "composite": true,
+        "declaration": true,
+        "forceConsistentCasingInFileNames": true
+    }
 }

